avm/utilities/pipelines/platform/Invoke-AvmJsonModuleIndexGeneration.ps1

```diff
@@ -0,0 +1,169 @@
+            Write-Output "Error message: $($_.Exception.Message)"
+            continue
+          }
+          $tags = $tagListResponse.tags | Sort-Object
+
+          $properties = [ordered]@{}
+          foreach ($tag in $tags) {
+            $gitTag = "$modulePath/$tag"
+            $documentationUri = "https://github.com/Azure/bicep-registry-modules/tree/$gitTag/$modulePath/README.md"
+
+            try {
+              $moduleMainJsonUri = "https://raw.githubusercontent.com/Azure/bicep-registry-modules/$gitTag/$mainJsonPath"
+              Write-Verbose "    Getting available description for tag $tag via '$moduleMainJsonUri'..." -Verbose
+              $moduleMainJsonUriResponse = Invoke-RestMethod -Uri $moduleMainJsonUri
+              $description = $moduleMainJsonUriResponse.metadata.description
+            } catch {
+              Write-Output "Error occurred while accessing description for tag $tag via '$moduleMainJsonUri'"
+              Write-Output "Error message: $($_.Exception.Message)"
+              continue
+            }
+
+            $properties[$tag] = [ordered]@{
+              'description'      = $description
+              'documentationUri' = $documentationUri
+            }
+          }
+
+          $moduleIndexData += [ordered]@{
+            'moduleName' = $modulePath
+            'tags'       = @($tags)
+            'properties' = $properties
+          }
+        } catch {
+          Write-Error "Error message: $($_.Exception.Message)"
+        }
+      }
+
+      $numberOfModuleGroupsProcessed++
+    }
+  }
+
+  Write-Verbose "Processed $numberOfModuleGroupsProcessed modules groups." -Verbose
+  Write-Verbose "Processed $($moduleIndexData.Count) total modules." -Verbose
+
+  Write-Verbose "Convert moduleIndexData variable to JSON and save as 'generated-moduleIndex.json'" -Verbose
+  $moduleIndexData | ConvertTo-Json -Depth 10 | Out-File -FilePath $currentGeneratedModuleIndexJsonFilePath
+
+  ## Merge the new moduleIndex.json file with the previous version if the $doNotMergeWithLastModuleIndexJsonFileVersion is set to $false
+
+  if ($doNotMergeWithLastModuleIndexJsonFileVersion -eq $false) {
+    Write-Verbose "Merging 'generated-moduleIndex.json' (new) file with 'last-moduleIndex.json' (previous) file..." -Verbose
+
+    $lastModuleIndexJsonFileContent = Get-Content $lastModuleIndexJsonFilePath
+    $currentGeneratedModuleIndexJsonFileContent = Get-Content $currentGeneratedModuleIndexJsonFilePath
+
+    $lastModuleIndexData = $lastModuleIndexJsonFileContent | ConvertFrom-Json -Depth 10
+    $currentGeneratedModuleIndexData = $currentGeneratedModuleIndexJsonFileContent | ConvertFrom-Json -Depth 10
+
+    $mergedModuleIndexData = $lastModuleIndexData + $currentGeneratedModuleIndexData | Group-Object -Property moduleName | ForEach-Object {
+      $_.Group | Sort-Object -Property @{Expression = { $_.tags -join ',' } } -Descending | Select-Object -First 1
+    } | Sort-Object -Property moduleName
```

Comment:
Just ran a quick test and it seems the above logic throws the `currentIndexData` away instead of merging the two:
```PowerShell
$hash1 = @(
    @{
        'moduleName' = 'bicep/avm/res/key-vault/vault'
        'tags'       = @(
            '0.3.0'
            '0.4.0'
            '0.5.0'
        )
    }
)
$hash2 = @(
    @{
        'moduleName' = 'bicep/avm/res/key-vault/vault'
        'tags'       = @(
            '0.1.0'
            '0.2.0'
            '0.3.0'
        )
    }
)


$mergedModuleIndexData = $hash1 + $hash2 | Group-Object -Property moduleName | ForEach-Object {
    $_.Group | Sort-Object -Property @{Expression = { $_.tags -join ',' } } -Descending | Select-Object -First 1
} | Sort-Object -Property moduleName

$mergedModuleIndexData

# Output
# $mergedModuleIndexData.tags
# 0.3.0
# 0.4.0
# 0.5.0
```

Reply:
It took some time, but this should work
```suggestion
    $mergedModuleIndexData = ($lastModuleIndexData + $currentGeneratedModuleIndexData) | Group-Object -Property 'moduleName' | ForEach-Object {
        $newTags = $_.Group.tags | Sort-Object -Unique
    
        $newProperties = [ordered]@{}
        $oldProperties = $_.Group.properties
        $oldProperties.Keys | Sort-Object | ForEach-Object {
            if ($newProperties.Keys -notContains $_) {
                $newProperties[$_] = $oldProperties[0].Keys -contains $_ ? $oldProperties[0][$_] : $oldProperties[1][$_]
            }
        }
    
        $_.Group | ForEach-Object {
            $_.tags = $newTags
            $_.properties = $newProperties
        }
        
        $_.Group | Select-Object -First 1
    }
```

Reply:
A test with the new code results in:

```PowerShell
$hash1 = @(
    @{
        'moduleName' = 'bicep/avm/res/key-vault/vault'
        'tags'       = @(
            '0.3.0'
            '0.4.0'
            '0.5.0'
        )
        'properties' = @{
            '0.3.0' = @{
                'description'      = 'desc1'
                'documentationUri' = 'doc1'
            }
            '0.4.0' = @{
                'description'      = 'desc2'
                'documentationUri' = 'doc2'
            }
            '0.5.0' = @{
                'description'      = 'desc3'
                'documentationUri' = 'doc3'
            }
        }
    }
    @{
        'moduleName' = 'bicep/avm/res/storage-account/account'
        'tags'       = @(
            '0.3.0'
            '0.4.0'
            '0.5.0'
        )
        'properties' = @{
            '0.3.0' = @{
                'description'      = 'desc3'
                'documentationUri' = 'doc3'
            }
            '0.4.0' = @{
                'description'      = 'desc4'
                'documentationUri' = 'doc4'
            }
            '0.5.0' = @{
                'description'      = 'desc5'
                'documentationUri' = 'doc5'
            }
        }
    }
)


$hash2 = @(
    @{
        'moduleName' = 'bicep/avm/res/key-vault/vault'
        'tags'       = @(
            '0.1.0'
            '0.2.0'
            '0.3.0'
        )
        'properties' = @{
            '0.1.0' = @{
                'description'      = 'desc1'
                'documentationUri' = 'doc1'
            }
            '0.2.0' = @{
                'description'      = 'desc2'
                'documentationUri' = 'doc2'
            }
            '0.3.0' = @{
                'description'      = 'desc3'
                'documentationUri' = 'doc3'
            }
        }
    }
    @{
        'moduleName' = 'bicep/avm/res/storage-account/account'
        'tags'       = @(
            '0.6.0'
            '0.8.0'
            '0.9.0'
        )
        'properties' = @{
            '0.6.0' = @{
                'description'      = 'desc6'
                'documentationUri' = 'doc6'
            }
            '0.8.0' = @{
                'description'      = 'desc8'
                'documentationUri' = 'doc8'
            }
            '0.9.0' = @{
                'description'      = 'desc9'
                'documentationUri' = 'doc9'
            }
        }
    }
)



$mergedModuleIndexData = ($hash1 + $hash2) | Group-Object -Property 'moduleName' | ForEach-Object {
    $newTags = $_.Group.tags | Sort-Object -Unique

    $newProperties = [ordered]@{}
    $oldProperties = $_.Group.properties
    $oldProperties.Keys | Sort-Object | ForEach-Object {
        if ($newProperties.Keys -notContains $_) {
            $newProperties[$_] = $oldProperties[0].Keys -contains $_ ? $oldProperties[0][$_] : $oldProperties[-1][$_]
        }
    }

    $_.Group | ForEach-Object {
        $_.tags = $newTags
        $_.properties = $newProperties
    }
    $_.Group | Select-Object -First 1
}

$mergedModuleIndexData
```

```json
[
  {
    "moduleName": "bicep/avm/res/key-vault/vault",
    "properties": {
      "0.1.0": {
        "description": "desc1",
        "documentationUri": "doc1"
      },
      "0.2.0": {
        "description": "desc2",
        "documentationUri": "doc2"
      },
      "0.3.0": {
        "description": "desc1",
        "documentationUri": "doc1"
      },
      "0.4.0": {
        "description": "desc2",
        "documentationUri": "doc2"
      },
      "0.5.0": {
        "description": "desc3",
        "documentationUri": "doc3"
      }
    },
    "tags": [
      "0.1.0",
      "0.2.0",
      "0.3.0",
      "0.4.0",
      "0.5.0"
    ]
  },
  {
    "moduleName": "bicep/avm/res/storage-account/account",
    "properties": {
      "0.3.0": {
        "description": "desc3",
        "documentationUri": "doc3"
      },
      "0.4.0": {
        "description": "desc4",
        "documentationUri": "doc4"
      },
      "0.5.0": {
        "description": "desc5",
        "documentationUri": "doc5"
      },
      "0.6.0": {
        "description": "desc6",
        "documentationUri": "doc6"
      },
      "0.8.0": {
        "description": "desc8",
        "documentationUri": "doc8"
      },
      "0.9.0": {
        "description": "desc9",
        "documentationUri": "doc9"
      }
    },
    "tags": [
      "0.3.0",
      "0.4.0",
      "0.5.0",
      "0.6.0",
      "0.8.0",
      "0.9.0"
    ]
  }
]
```